### DISCUSSION TITLE
二分搜尋法實作-迴圈版本

### DISCUSSION THREAD

**Discussion by eric202501** [1101-GS4719-C/Python-Community-GS4719-C#1624] [General]
二分搜尋法為眾多演算法中基礎的一環，前題條件為在一排列好的數列中(小至大)搜尋目標。
若搜尋到目標，輸出其index值。若沒有查詢到目標值，則輸出目標不在陣列中

程式實作
<img width="307" alt="image" src="https://github.com/NCU-GS4719-Python/Python-Community/assets/144879605/5a93a480-3a56-4789-9edc-93e07e252f9f">
<img width="245" alt="image" src="https://github.com/NCU-GS4719-Python/Python-Community/assets/144879605/19e3f008-e355-434e-92ce-d3ab4e482ce0">
<img width="143" alt="image" src="https://github.com/NCU-GS4719-Python/Python-Community/assets/144879605/9e702714-2a62-4f8d-82b7-753ae3a141a3">
首先，我們在陣列的兩側設置左邊界和右邊界(圖為參考) 
以下簡稱左邊界為Left 右亦然 只要left比right小 意味著邊界正常 就持續執行搜尋

<img width="299" alt="image" src="https://github.com/NCU-GS4719-Python/Python-Community/assets/144879605/b77758f1-691a-44b8-8ca1-2aca8163cd93">

mid = 左右邊界的中間點的index 
如果target>array[mid]意味著我們的目標在mid的右方 故mid的左方都可以不用看了  左邊界設為mid+1
如果target<array[mid]意味著我們的目標在mid的左方 故mid的右方都可以不用看了  右邊界設為mid-1
一旦target=array[mid] 代表我們找到我們的目標，就可以輸出我們的index了
若邊界執行至極限(使左右邊界相交)卻尚未return index的話 就是該值不存在於array中了
當然，二分搜還有許多特例或針對特殊的限制，例如一旦陣列中存在重複的數值該怎麼辦?
下次再說

**Comment by lazywulf** [comment]
我建議版主寫 Pseudocode 而非直接將程式碼貼出。雖然 python 語法近似英文，簡單易懂，不過對於不熟悉 python 的人而言會有些困難。
另外，對於所介紹的演算法，希望版主能給出時間與空間複雜度的分析。雖然對此例而言算是簡單瑣碎的 (Time complexity: O(log(n))), 不過換成其他例子的話給出這樣的分析是很有幫助的喔。

**Comment by ATREE01** [comment]
這邊建議如果要貼程式碼的話可以使用github 內建的功能 前後用三個 \` 夾住程式碼
\`\`\`python 
print("Hello World")
\`\`\`
這樣打的話就會變成以下這樣，還會幫你上顏色，美觀程度大提升。
```python
print("Hello World")
```

**Reply by 1chooo** [reply]
沒錯！這的確是非常好的建議！感謝你的分享！我們通常在 GitHub 上寫文檔的時後，附上程式邏輯時，也都是透過這樣的方法為程式碼加上 highlight。
